source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -10,709 +10,1371 @@
+                    else if (auto layoutDecor = key->findDecoration<IRLayoutDecoration>())
+                    {
+                        layoutDecor->removeAndDeallocate();
+                        continue;
+                    }
+                    break;
+                }
             }
         }
 
-        IRBuilder builder(func);
-        builder.setInsertBefore(func);
-        for (auto param : paramsToProcess)
+        void hoistEntryPointParameterFromStruct(EntryPointInfo entryPoint)
         {
-            auto structType = as<IRStructType>(param->getDataType());
-            builder.setInsertBefore(func->getFirstBlock()->getFirstOrdinaryInst());
-            auto varLayout = findVarLayout(param);
-
-            // If `param` already has a semantic, we don't want to hoist its fields out.
-            if (varLayout->findSystemValueSemanticAttr() != nullptr ||
-                param->findDecoration<IRSemanticDecoration>())
-                continue;
-
-            IRStructTypeLayout* structTypeLayout = nullptr;
-            if (varLayout)
-                structTypeLayout = as<IRStructTypeLayout>(varLayout->getTypeLayout());
-            Index fieldIndex = 0;
-            List<IRInst*> fieldParams;
-            for (auto field : structType->getFields())
+            // If an entry point has a input parameter with a struct type, we want to hoist out
+            // all the fields of the struct type to be individual parameters of the entry point.
+            // This will canonicalize the entry point signature, so we can handle all cases uniformly.
+
+            // For example, given an entry point:
+            // ```
+            // struct VertexInput { float3 pos; float 2 uv; int vertexId : SV_VertexID};
+            // void main(VertexInput vin) { ... }
+            // ```
+            // We will transform it to:
+            // ```
+            // void main(float3 pos, float2 uv, int vertexId : SV_VertexID) {
+            //     VertexInput vin = {pos,uv,vertexId};
+            //     ...
+            // }
+            // ```
+
+            auto func = entryPoint.entryPointFunc;
+            List<IRParam*> paramsToProcess;
+            for (auto param : func->getParams())
             {
-                auto fieldParam = builder.emitParam(field->getFieldType());
-                
-                IRCloneEnv cloneEnv;
-                cloneInstDecorationsAndChildren(&cloneEnv, builder.getModule(), field->getKey(), fieldParam);
+                if (as<IRStructType>(param->getDataType()))
+                {
+                    paramsToProcess.add(param);
+                }
+            }
 
-                IRVarLayout* fieldLayout = structTypeLayout ? structTypeLayout->getFieldLayout(fieldIndex) : nullptr;
+            IRBuilder builder(func);
+            builder.setInsertBefore(func);
+            for (auto param : paramsToProcess)
+            {
+                auto structType = as<IRStructType>(param->getDataType());
```

Comment:
Should comment here that the struct type is already flattened earlier so we don't need to account for nested struct type fields here.

Reply:
The variable is not flattened by this pass: first we hoist, then pack `[StageIn]`, then reformat the `[StageIn]` to be legal

The following code demonstrates this
```
            //Input Parameter Legalize
            hoistEntryPointParameterFromStruct(entryPoint);
            packStageInParameters(entryPoint);
            flattenInputParameters(entryPoint);
```